[office] Add a shape visual for each page that has not been yet rendered.

dcaliste opened the pull request:
This is a small modification in the paint node method to display a diffused rectangle shape of a page when the page has not been rendered yet. This gives a minimal visual feedback of the position in the document for instance when scrolling fast, or when  going to the next search match several pages further.

pvuorela commented on pdf/pdfcanvas.cpp:
```diff
@@ -398,6 +398,10 @@ QSGNode* PDFCanvas::updatePaintNode(QSGNode* node, QQuickItem::UpdatePaintNodeDa
         }
+        else if( t->childCount() > 0 ) {
+            delete t->firstChild();
+            t->removeAllChildNodes();
```
This would delete and below recreate child nodes when view is scrolled even a bit. Should avoid if it's already the wanted type of node?

dcaliste replied:
Yes, I see. I'll rework this.

pvuorela commented on pdf/pdfcanvas.cpp:
```diff
@@ -449,13 +453,12 @@ QSGNode* PDFCanvas::updatePaintNode(QSGNode* node, QQuickItem::UpdatePaintNodeDa
+            QSGSimpleRectNode *bgNode = new QSGSimpleRectNode();
+            t->appendChildNode( bgNode );
+            bgNode->setRect( QRectF(0., 0., page.rect.width(), page.rect.height()) );
+            bgNode->setColor( d->linkColor );
```
Rectangle is not really a link, could define a separate property perhaps?

dcaliste replied:
I agree, I was a bit lazy last Sunday and quite satisfied with the visual result because the link color property is set to Theme.highlightColor plus an alpha… I'll add one specific property which is cleaner than supposing that the link color will always be highlightColor !

dcaliste commented:
I've added a pagePlaceholderColor property. Maybe the name is too long ?

I've changed a bit the paint function to avoid deleting the QGSNode each time we paint. I didn't find how to differentiate QGSSimpleTextureNode and QSGSimpleRectangleNode because they are not inheriting from QObject so qobject_cast() is not available. Instead I've added a flag (rectPlaceholder) that is true when the first child of the QGSNode is indeed a rectangle, to know if we need to delete the node to insert the texture. What do you think ?

pvuorela commented on pdf/pdfcanvas.cpp:
```diff
@@ -34,12 +34,13 @@ struct PDFPage {
     int index;
     QRectF rect;
-    bool requested;
+    bool requested, rectPlaceholder;
```
As personal preference, I'd keep member variable declarations separate to ease reading and follow existing convention. Doesn't matter much, though.

dcaliste replied:
You're right. Commit changed.

pvuorela commented on pdf/pdfcanvas.cpp:
```diff
@@ -226,6 +243,7 @@ void PDFCanvas::layout()
         page.index = i;
         page.rect = QRectF(0, totalHeight, width(), width() * ratio);
         page.links = links.values( i );
+        page.rectPlaceholder = false;
```
Minor detail, but looks unnecessary, like line below too, as ctor sets false already.

dcaliste replied:
I agree, that's right. Commit changed.

pvuorela commented:
I think the name is fine. Not referenced in many places so clarity wins over few characters.

Couple cosmetic comments, which don't matter much. LGTM.

dcaliste commented:
I've corrected the minor details by the way.

pvuorela commented:
Looking good. Thanks!
